StardewArchipelago/Locations/CodeInjections/Modded/MagicModInjections.cs

```diff
@@ -0,0 +1,228 @@
+                            if (index == 395) // Coffee
+                                spellsLearnedList.Add(LEARN_HASTE_AP_LOCATION);
+                            else if (index == 773) // Life elixir
+                                spellsLearnedList.Add(LEARN_HEAL_AP_LOCATION);
+                            else if (index == 86) // Earth crystal
+                                spellsLearnedList.Add(LEARN_SHOCKWAVE_AP_LOCATION);
+                            else if (index == 82) // Fire quartz
+                                spellsLearnedList.Add(LEARN_FIREBALL_AP_LOCATION);
+                            else if (index == 161) // Ice Pip
+                                spellsLearnedList.Add(LEARN_FROSTBITE_AP_LOCATION);
+                        }
+                    }
+                }
+                foreach (var lightSource in player.currentLocation.sharedLights.Values)
+                {
+                    if (Utility.distance(targetX, lightSource.position.X, targetY, lightSource.position.Y) < lightSource.radius.Value * Game1.tileSize)
+                    {
+                        spellsLearnedList.Add(LEARN_LANTERN_AP_LOCATION);
+                        break;
+                    }
+                }
+                var tilePos = new Vector2(targetX / Game1.tileSize, targetY / Game1.tileSize);
+                if (player.currentLocation.terrainFeatures.ContainsKey(tilePos) && player.currentLocation.terrainFeatures[tilePos] is StardewValley.TerrainFeatures.HoeDirt hd)
+                {
+                    if (hd.crop != null)
+                        spellsLearnedList.Add(LEARN_TENDRILS_AP_LOCATION);
+                }
+                var tile = player.currentLocation.map.GetLayer("Buildings").Tiles[(int)tilePos.X, (int)tilePos.Y];
+                if (tile != null && tile.TileIndex == 173)
+                    spellsLearnedList.Add(LEARN_DESCEND_AP_LOCATION);
+                if (player.currentLocation is Farm farm)
+                {
+                    foreach (var clump in farm.resourceClumps)
+                    {
+                        if (clump.parentSheetIndex.Value == 622 && new Rectangle((int)clump.tile.Value.X, (int)clump.tile.Value.Y, clump.width.Value, clump.height.Value).Contains((int)tilePos.X, (int)tilePos.Y))
+                            spellsLearnedList.Add(LEARN_METEOR_AP_LOCATION);
+                    }
+                }
+                if (player.currentLocation.doesTileHaveProperty((int)tilePos.X, (int)tilePos.Y, "Action", "Buildings") == "EvilShrineLeft")
+                    spellsLearnedList.Add(LEARN_LUCKSTEAL_AP_LOCATION);
+                if (player.currentLocation is StardewValley.Locations.MineShaft ms && ms.mineLevel == 100 && ms.waterTiles[(int)tilePos.X, (int)tilePos.Y])
+                    spellsLearnedList.Add(LEARN_BLOODMANA_AP_LOCATION);
+
+                for (int i = spellsLearnedList.Count - 1; i >= 0; --i)
```

Comment:
What is the point of this? Rechecking locations is fine?

Reply:
If the count is bigger than zero there's a little chime that plays that signals to the player that they found something normally.  If its already checked it won't play.

Reply:
Then you don't need to mess with the collection, just go
```
if (spellsLearnedList.Any(spell => _locationChecker.IsLocationMissing(spell)))
{
    Game1.playSound("secret1");
}
```